{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6bea81bb-5517-48a7-98f4-2c55eb7e0c1b",
   "metadata": {},
   "source": [
    "# LINGER ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cc667cf-0870-4cd8-b15c-dd6ba8062ebb",
   "metadata": {},
   "source": [
    "## Preparation steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "af4dd3e8-950e-4f2e-aaa9-eb499bf95c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "os.chdir('/home/kl467102/thesis/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4e9e06c3-69b2-4de5-939f-af3488915cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scipy\n",
    "import pandas as pd\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "245bf1b3-5696-4a9d-9f27-30b35b7c7c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/home/kl467102/thesis/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e1d6f02-ed13-4dec-80b1-8963360c3852",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '10x_output_hESC' # path to 10x output files\n",
    "label_file = '10x_output_hESC/label.txt' # path to cell-type annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af75bf70-b734-4a9f-83a4-5b4db898f8a0",
   "metadata": {},
   "source": [
    "The input consists of `matrix.mtx`, `features.tsv` and `barcodes.tsv` and an annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d0da79b9-8c59-4199-b201-c9d3d4647198",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix=scipy.io.mmread(os.path.join(input_dir, 'matrix.mtx'))\n",
    "features=pd.read_csv(os.path.join(input_dir, 'features.tsv'), sep='\\t', header=None)\n",
    "barcodes=pd.read_csv(os.path.join(input_dir, 'barcodes.tsv'), sep='\\t', header=None)\n",
    "label=pd.read_csv(label_file, sep='\\t', header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52acc68-f1e0-4837-b271-9b06f6162871",
   "metadata": {},
   "source": [
    "We have loaded the data, now we can inspect them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8dd76a2d-876f-461b-bc60-d6fd5c995848",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17736, 758)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3c4592b4-c2b8-4c7a-a466-0b4c1b8c80f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[2.5101334 , 0.        , 1.70225118, 4.25937163, 2.85114345],\n",
       "       [1.42489576, 1.20869914, 0.        , 0.        , 0.        ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.        ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.        ],\n",
       "       [2.14110522, 2.93625025, 0.22070073, 0.77659877, 1.37593925]])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dense_matrix = matrix.toarray()\n",
    "subset = dense_matrix[:5, :5]\n",
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b471ef9-31e0-4850-bc6a-6a922b8fe916",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17734</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17735</th>\n",
       "      <td>KI270713.1:31340-32243</td>\n",
       "      <td>KI270713.1:31340-32243</td>\n",
       "      <td>Peaks</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            0                       1                2\n",
       "0                        A1BG                    A1BG  Gene Expression\n",
       "1                        A1CF                    A1CF  Gene Expression\n",
       "17734                    ZZZ3                    ZZZ3  Gene Expression\n",
       "17735  KI270713.1:31340-32243  KI270713.1:31340-32243            Peaks"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([features.head(2), features.tail(2)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "39dc8531-8d7f-4db9-b428-a75ad0c6ccea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCTCGCTTCATCTTCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGGGGTGGGCGTTAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTCGTATAAGTATCAT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0\n",
       "0  TCTCGCTTCATCTTCA-1\n",
       "1  CGGGGTGGGCGTTAAC-1\n",
       "2  TTCGTATAAGTATCAT-1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "49a27e9c-94e3-4e9d-86e0-0286d5946698",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode_use</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGGGGTGGGCGTTAAC-1</td>\n",
       "      <td>test cell type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTCGTATAAGTATCAT-1</td>\n",
       "      <td>test cell type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTACGTGGCGTGACCC-1</td>\n",
       "      <td>test cell type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          barcode_use           label\n",
       "0  CGGGGTGGGCGTTAAC-1  test cell type\n",
       "1  TTCGTATAAGTATCAT-1  test cell type\n",
       "2  CTACGTGGCGTGACCC-1  test cell type"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a42c28-459c-45b4-9a23-32c918decaf3",
   "metadata": {},
   "source": [
    "## INPUT MANIPULATION - we need to make it work"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a20adef6-ea94-4cbd-b409-97dab2ab4e27",
   "metadata": {},
   "source": [
    "In this part, we test LINGER performance on dataset provided in the BEELINE, namely hESC data (scRNA-Seq on human embrionic stem cells)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e666acd-be9d-4084-aff1-8833a17ab9b0",
   "metadata": {},
   "source": [
    "We need to convert provided data in `ExpressionData.csv` to format accepted by LINGER, i.e. generate `matrix.mtx`, `features.csv` and `barcodes.tsv`. We also create mock `label.txt`. Chromosome location needs to be updated based on annotation file for `features.tsv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "308d7bd7-852f-4bb0-9eec-2b2e4a87bc20",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/home/kl467102/thesis/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "74445104-3259-42a3-acf2-d90cf4a76d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def generate_atac_for_rna(input_csv, features, num_regions = 500, out_csv = None):\n",
    "    # List of all chromosomes (Human Genome)\n",
    "    df = pd.read_csv(input_csv, index_col=0)\n",
    "    if not out_csv:\n",
    "        # Split the file path into directory, base name, and extension\n",
    "        directory, file_name = os.path.split(input_csv)\n",
    "        base_name, ext = os.path.splitext(file_name)\n",
    "        new_file_name = f\"{base_name}_modified{ext}\"\n",
    "        out_csv = os.path.join(directory, new_file_name)\n",
    "\n",
    "    chromosomes = [\n",
    "        'chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', \n",
    "        'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', \n",
    "        'chr20', 'chr21', 'chr22', 'chrX', 'chrY'\n",
    "    ]\n",
    "\n",
    "    # Chromosome lengths (based on the human genome assembly GRCh38)\n",
    "    max_chrom_length = {\n",
    "        'chr1': 249250621, 'chr2': 243199373, 'chr3': 198022430, 'chr4': 191154276, \n",
    "        'chr5': 180915260, 'chr6': 171115067, 'chr7': 159138663, 'chr8': 146364022, \n",
    "        'chr9': 141213431, 'chr10': 135534747, 'chr11': 135006516, 'chr12': 133851895, \n",
    "        'chr13': 115169878, 'chr14': 107349540, 'chr15': 102531392, 'chr16': 90354753, \n",
    "        'chr17': 81195210, 'chr18': 78077248, 'chr19': 59128983, 'chr20': 63025520, \n",
    "        'chr21': 48129895, 'chr22': 50818468, 'chrX': 156040895, 'chrY': 57227415\n",
    "    }\n",
    "    new_df = df.copy()\n",
    "    new_features = features.copy()\n",
    "    num_samples = new_df.shape[1]\n",
    "    # Generate random genomic regions\n",
    "    regions = []\n",
    "    region_index = []\n",
    "    \n",
    "    for _ in range(num_regions):\n",
    "        chrom = np.random.choice(chromosomes)\n",
    "        start_pos = np.random.randint(0, max_chrom_length[chrom] - 1000)  # random start position\n",
    "        end_pos = start_pos + np.random.randint(500, 1000)  # random length between 500 and 1000\n",
    "        feature = f'{chrom}:{start_pos}-{end_pos}'\n",
    "        region_index.append(feature)\n",
    "        regions.append([feature, chrom, start_pos, end_pos])\n",
    "        #new_features.loc[len(new_features)] = []\n",
    "    # Generate random values for ATAC-seq signal (e.g., counts)\n",
    "    #random_values = np.random.poisson(lam=5, size=(num_regions, num_samples))  # Poisson distribution for count data\n",
    "    theta = 1\n",
    "    mu = 0.5\n",
    "    random_values = np.random.negative_binomial(n=theta, p=theta / (theta + mu), size=(num_regions, num_samples))\n",
    "\n",
    "    # Create a DataFrame for the random values\n",
    "    random_atac_df = pd.DataFrame(random_values, columns=new_df.columns)\n",
    "    random_atac_df.index = region_index\n",
    "    # Concatenate the random ATAC-seq data with the existing features DataFrame\n",
    "    #updated_features_df = pd.concat([features_df, random_atac_df], axis=1)\n",
    "\n",
    "    # Create a DataFrame\n",
    "    #regions_df = pd.DataFrame(regions, columns=[\"feature\", \"Chromosome\", \"Start\", \"End\"])\n",
    "\n",
    "    # Combine the data into a final DataFrame\n",
    "    #final_df = pd.concat([regions_df, random_data_df], axis=1)\n",
    "    new_expr = pd.concat([df, random_atac_df])\n",
    "    # Save to CSV (optional)\n",
    "    new_expr.to_csv(out_csv, index=True)\n",
    "    return new_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "5f2f9611-e084-404f-8ddc-9dbc48e644dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_csv = \"data/hESC_ExpressionData.csv\"  # Input CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "3f055502-a904-482b-b880-f50d319af6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = generate_atac_for_rna(input_csv, features, num_regions = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e85052e1-bd83-4072-b6b3-600edc653a7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H9_00hb4s_001</th>\n",
       "      <th>H9_00hb4s_002</th>\n",
       "      <th>H9_00hb4s_003</th>\n",
       "      <th>H9_00hb4s_004</th>\n",
       "      <th>H9_00hb4s_005</th>\n",
       "      <th>H9_00hb4s_006</th>\n",
       "      <th>H9_00hb4s_007</th>\n",
       "      <th>H9_00hb4s_008</th>\n",
       "      <th>H9_00hb4s_009</th>\n",
       "      <th>H9_00hb4s_010</th>\n",
       "      <th>...</th>\n",
       "      <th>H9_96h_182</th>\n",
       "      <th>H9_96h_183</th>\n",
       "      <th>H9_96h_184</th>\n",
       "      <th>H9_96h_185</th>\n",
       "      <th>H9_96h_186</th>\n",
       "      <th>H9_96h_187</th>\n",
       "      <th>H9_96h_188</th>\n",
       "      <th>H9_96h_189</th>\n",
       "      <th>H9_96h_190</th>\n",
       "      <th>H9_96h_192</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.510133</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.702251</td>\n",
       "      <td>4.259372</td>\n",
       "      <td>2.851143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.357210</td>\n",
       "      <td>0.925952</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.473057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.946191</td>\n",
       "      <td>3.524386</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.516034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.510372</td>\n",
       "      <td>1.418294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>1.424896</td>\n",
       "      <td>1.208699</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.130935</td>\n",
       "      <td>1.880416</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.910541</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.804444</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.883394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.295367</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>2.141105</td>\n",
       "      <td>2.936250</td>\n",
       "      <td>0.220701</td>\n",
       "      <td>0.776599</td>\n",
       "      <td>1.375939</td>\n",
       "      <td>1.473409</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.359460</td>\n",
       "      <td>2.312497</td>\n",
       "      <td>1.584052</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.244709</td>\n",
       "      <td>1.886233</td>\n",
       "      <td>0.632573</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.195413</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.539408</td>\n",
       "      <td>3.034209</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:23615416-23616071</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr8:113571844-113572367</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr14:86401053-86401586</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21:2429270-2430038</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13:112148230-112149130</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18235 rows × 758 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           H9_00hb4s_001  H9_00hb4s_002  H9_00hb4s_003  \\\n",
       "A1BG                            2.510133       0.000000       1.702251   \n",
       "A1CF                            1.424896       1.208699       0.000000   \n",
       "A2LD1                           0.000000       0.000000       0.000000   \n",
       "A2M                             0.000000       0.000000       0.000000   \n",
       "A2ML1                           2.141105       2.936250       0.220701   \n",
       "...                                  ...            ...            ...   \n",
       "chrY:23615416-23616071          3.000000       0.000000       1.000000   \n",
       "chr8:113571844-113572367        1.000000       2.000000       0.000000   \n",
       "chr14:86401053-86401586         3.000000       0.000000       1.000000   \n",
       "chr21:2429270-2430038           0.000000       0.000000       1.000000   \n",
       "chr13:112148230-112149130       1.000000       0.000000       0.000000   \n",
       "\n",
       "                           H9_00hb4s_004  H9_00hb4s_005  H9_00hb4s_006  \\\n",
       "A1BG                            4.259372       2.851143       0.000000   \n",
       "A1CF                            0.000000       0.000000       0.000000   \n",
       "A2LD1                           0.000000       0.000000       0.000000   \n",
       "A2M                             0.000000       0.000000       0.000000   \n",
       "A2ML1                           0.776599       1.375939       1.473409   \n",
       "...                                  ...            ...            ...   \n",
       "chrY:23615416-23616071          0.000000       0.000000       0.000000   \n",
       "chr8:113571844-113572367        2.000000       1.000000       3.000000   \n",
       "chr14:86401053-86401586         0.000000       0.000000       0.000000   \n",
       "chr21:2429270-2430038           1.000000       0.000000       0.000000   \n",
       "chr13:112148230-112149130       0.000000       0.000000       0.000000   \n",
       "\n",
       "                           H9_00hb4s_007  H9_00hb4s_008  H9_00hb4s_009  \\\n",
       "A1BG                            2.357210       0.925952       0.000000   \n",
       "A1CF                            1.130935       1.880416       0.000000   \n",
       "A2LD1                           0.000000       0.000000       2.295367   \n",
       "A2M                             0.000000       0.000000       0.000000   \n",
       "A2ML1                           0.000000       1.359460       2.312497   \n",
       "...                                  ...            ...            ...   \n",
       "chrY:23615416-23616071          0.000000       0.000000       0.000000   \n",
       "chr8:113571844-113572367        0.000000       1.000000       0.000000   \n",
       "chr14:86401053-86401586         0.000000       1.000000       2.000000   \n",
       "chr21:2429270-2430038           1.000000       0.000000       0.000000   \n",
       "chr13:112148230-112149130       0.000000       1.000000       0.000000   \n",
       "\n",
       "                           H9_00hb4s_010  ...  H9_96h_182  H9_96h_183  \\\n",
       "A1BG                            1.473057  ...         0.0    1.946191   \n",
       "A1CF                            0.000000  ...         0.0    0.000000   \n",
       "A2LD1                           0.000000  ...         0.0    0.000000   \n",
       "A2M                             0.000000  ...         0.0    0.000000   \n",
       "A2ML1                           1.584052  ...         0.0    1.244709   \n",
       "...                                  ...  ...         ...         ...   \n",
       "chrY:23615416-23616071          0.000000  ...         0.0    0.000000   \n",
       "chr8:113571844-113572367        1.000000  ...         0.0    0.000000   \n",
       "chr14:86401053-86401586         1.000000  ...         0.0    1.000000   \n",
       "chr21:2429270-2430038           0.000000  ...         0.0    1.000000   \n",
       "chr13:112148230-112149130       0.000000  ...         0.0    0.000000   \n",
       "\n",
       "                           H9_96h_184  H9_96h_185  H9_96h_186  H9_96h_187  \\\n",
       "A1BG                         3.524386    0.000000         0.0    3.516034   \n",
       "A1CF                         0.000000    1.910541         0.0    0.000000   \n",
       "A2LD1                        0.000000    0.000000         0.0    0.000000   \n",
       "A2M                          0.000000    0.000000         0.0    0.000000   \n",
       "A2ML1                        1.886233    0.632573         0.0    1.195413   \n",
       "...                               ...         ...         ...         ...   \n",
       "chrY:23615416-23616071       1.000000    0.000000         1.0    0.000000   \n",
       "chr8:113571844-113572367     1.000000    0.000000         1.0    0.000000   \n",
       "chr14:86401053-86401586      0.000000    0.000000         0.0    0.000000   \n",
       "chr21:2429270-2430038        0.000000    1.000000         2.0    0.000000   \n",
       "chr13:112148230-112149130    0.000000    0.000000         1.0    0.000000   \n",
       "\n",
       "                           H9_96h_188  H9_96h_189  H9_96h_190  H9_96h_192  \n",
       "A1BG                              0.0    0.000000    1.510372    1.418294  \n",
       "A1CF                              0.0    0.804444    0.000000    0.883394  \n",
       "A2LD1                             0.0    0.000000    0.000000    0.000000  \n",
       "A2M                               0.0    0.000000    0.000000    0.000000  \n",
       "A2ML1                             0.0    1.539408    3.034209    0.000000  \n",
       "...                               ...         ...         ...         ...  \n",
       "chrY:23615416-23616071            1.0    3.000000    0.000000    0.000000  \n",
       "chr8:113571844-113572367          1.0    2.000000    2.000000    2.000000  \n",
       "chr14:86401053-86401586           0.0    0.000000    0.000000    0.000000  \n",
       "chr21:2429270-2430038             0.0    0.000000    1.000000    0.000000  \n",
       "chr13:112148230-112149130         1.0    3.000000    1.000000    0.000000  \n",
       "\n",
       "[18235 rows x 758 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc6c76bf-cc0e-4889-a0d9-af194d700b08",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.io import mmwrite\n",
    "from scipy.sparse import csr_matrix\n",
    "import random\n",
    "\n",
    "\n",
    "def generate_sample_barcodes(num_barcodes, length=16):\n",
    "    \"\"\"Generate unique barcodes for a single sample, all prefixed with '1-'.\"\"\"\n",
    "    barcodes = set()\n",
    "    nucleotides = \"ATCG\"\n",
    "    while len(barcodes) < num_barcodes:\n",
    "        random_sequence = \"\".join(random.choices(nucleotides, k=length))\n",
    "        barcode = f\"{random_sequence}-1\"\n",
    "        barcodes.add(barcode)\n",
    "    return list(barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "95e62d8c-05b4-4b77-ae6f-d1c950f25b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import os\n",
    "import re\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy.io import mmwrite\n",
    "\n",
    "def process_csv_to_10x(input_csv, output_dir, barcode_length=16):\n",
    "    \"\"\"\n",
    "    Processes input CSV to 10x-compatible output files, handling gene and chromatin accessibility data.\n",
    "\n",
    "    Args:\n",
    "        input_csv (str): Path to input CSV file.\n",
    "        output_dir (str): Directory to save 10x output files.\n",
    "        barcode_length (int): Length of barcodes to generate for samples.\n",
    "    \"\"\"\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "\n",
    "    # Read the input CSV file\n",
    "    data = pd.read_csv(input_csv, index_col=0)\n",
    "\n",
    "    # Extract row identifiers\n",
    "    row_ids = data.index.tolist()\n",
    "    num_samples = data.shape[1]  # Number of columns = number of samples\n",
    "\n",
    "    # Classify rows: genes vs chromatin accessibility\n",
    "    gene_rows = []\n",
    "    chromatin_rows = []\n",
    "\n",
    "    chromatin_pattern = re.compile(r\"^(chr|KI|GL)\\d+:\\d+-\\d+$\")  # Matches chromatin-like rows (e.g., chr1:1000-2000)\n",
    "    for row_id in row_ids:\n",
    "        if chromatin_pattern.match(row_id):\n",
    "            chromatin_rows.append(row_id)\n",
    "        else:\n",
    "            gene_rows.append(row_id)\n",
    "\n",
    "    # Generate barcodes for all samples\n",
    "    barcodes = generate_sample_barcodes(num_samples, length=barcode_length)\n",
    "\n",
    "    # Write features.tsv\n",
    "    features_file = f\"{output_dir}/features.tsv\"\n",
    "    with open(features_file, \"w\", newline=\"\") as f:\n",
    "        writer = csv.writer(f, delimiter=\"\\t\")\n",
    "        for gene in gene_rows:\n",
    "            writer.writerow([gene, gene, \"Gene Expression\"])\n",
    "        for chromatin in chromatin_rows:\n",
    "            writer.writerow([chromatin, chromatin, \"Peaks\"])\n",
    "\n",
    "    # Write barcodes.tsv\n",
    "    barcodes_file = f\"{output_dir}/barcodes.tsv\"\n",
    "    with open(barcodes_file, \"w\", newline=\"\") as f:\n",
    "        f.write(\"\\n\".join(barcodes) + \"\\n\")\n",
    "    \n",
    "    print(f\"Matrix shape before adding row: {data.shape}\")\n",
    "    # Add a new row with 0s for all columns\n",
    "    new_row = {column_name: 0 for column_name in data.columns}\n",
    "    data.loc['KI270713.1:31340-32243'] = new_row\n",
    "    print(f\"Matrix shape after adding row: {data.shape}\")\n",
    "\n",
    "    # Create a sparse matrix and write matrix.mtx\n",
    "    matrix_data = csr_matrix(data.values)\n",
    "    matrix_file = f\"{output_dir}/matrix.mtx\"\n",
    "    mmwrite(matrix_file, matrix_data)\n",
    "\n",
    "    print(\n",
    "        f\"Files created:\\n\"\n",
    "        f\"- {features_file}, number of features: {len(row_ids)} \"\n",
    "        f\"(Genes: {len(gene_rows)}, Accessibility: {len(chromatin_rows)})\\n\"\n",
    "        f\"- {barcodes_file}\\n\"\n",
    "        f\"- {matrix_file}, matrix shape: {matrix_data.shape}\"\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "3ef93015-de98-4e26-b67e-0687c2dcf7fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### OLD VERSION KEPT JUST IN CASE, NEED TO WYPIERDOLIC SOON\n",
    "'''\n",
    "def process_csv_to_10x(input_csv, output_dir, barcode_length=16):\n",
    "    # Read the input CSV file\n",
    "    data = pd.read_csv(input_csv, index_col=0)\n",
    "\n",
    "    # Extract gene names and sample names\n",
    "    gene_names = data.index.tolist()\t\n",
    "    num_samples = data.shape[1]  # Number of columns = number of samples\n",
    "\n",
    "    # Generate barcodes for all samples\n",
    "    barcodes = generate_sample_barcodes(num_samples, length=barcode_length)\n",
    "\n",
    "    # Write features.tsv (gene names)\n",
    "    features_file = f\"{output_dir}/features.tsv\"\n",
    "    with open(features_file, \"w\", newline=\"\") as f:\n",
    "        writer = csv.writer(f, delimiter=\"\\t\")\n",
    "        for gene in gene_names:\n",
    "            writer.writerow([gene, gene, \"Gene Expression\"])  # Example: gene_id, gene_name, feature_type\n",
    "    # Write barcodes.tsv\n",
    "    barcodes_file = f\"{output_dir}/barcodes.tsv\"\n",
    "    with open(barcodes_file, \"w\", newline=\"\") as f:\n",
    "        f.write(\"\\n\".join(barcodes) + \"\\n\")\n",
    "   \n",
    "    print(f'teraz mtx ma kształ {data.shape}')\n",
    "    # New row to add with 0s for all columns\n",
    "    new_row = {column_name: 0 for column_name in data.columns}\n",
    "    \n",
    "    # Add row using loc with a string index\n",
    "    data.loc['KI270713.1:31340-32243'] = new_row\n",
    "    print(f'po dodaniu ma ksztalt{data.shape}')\n",
    "    \n",
    "    # Create a sparse matrix and write matrix.mtx\n",
    "    matrix_data = csr_matrix(data.values)\n",
    "    matrix_file = f\"{output_dir}/matrix.mtx\"\n",
    "    mmwrite(matrix_file, matrix_data)\n",
    "\n",
    "    print(f\"Files created:\\n- {features_file}, number of features: {len(gene_names)}\\n- {barcodes_file}\\n- {matrix_file}, matrix shape: {matrix_data.shape}\")\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2f128c89-1d87-4f13-bc31-7dac3c144ec8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "teraz mtx ma kształ (17735, 758)\n",
      "po dodaniu ma ksztalt(17736, 758)\n",
      "Files created:\n",
      "- 10x_output_hESC/features.tsv, number of features: 17735\n",
      "- 10x_output_hESC/barcodes.tsv\n",
      "- 10x_output_hESC/matrix.mtx, matrix shape: (17736, 758)\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "input_csv = \"data/hESC_ExpressionData.csv\"  # Input CSV file\n",
    "output_dir = \"10x_output_hESC\"  # Directory to save output files\n",
    "\n",
    "import os\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "process_csv_to_10x(input_csv, output_dir, barcode_length=16)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "9e91b227-0108-4917-8c3c-e0b5873a6829",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Label file created: 10x_output_hESC/label.txt\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def create_label_file(input_file, output_file, label=\"test cell type\"):\n",
    "    \"\"\"\n",
    "    Create a label file with the specified label for all barcodes.\n",
    "    \n",
    "    Args:\n",
    "        input_file (str): Path to the input CSV file with barcode information.\n",
    "        output_file (str): Path to save the generated label.txt file.\n",
    "        label (str): The label to assign to all barcodes in the output file.\n",
    "    \"\"\"\n",
    "    # Read the input CSV file\n",
    "    data = pd.read_csv(input_file, sep=\"\\t\")\n",
    "    data.columns = ['barcode_use']\n",
    "    # Replace all labels with the provided label\n",
    "    data['label'] = label\n",
    "    \n",
    "    # Save the modified dataframe to the output file\n",
    "    data.to_csv(output_file, sep=\"\\t\", index=False, header=True, columns=[\"barcode_use\", \"label\"])\n",
    "    print(f\"Label file created: {output_file}\")\n",
    "\n",
    "\n",
    "# Example usage\n",
    "input_file = \"10x_output_hESC/barcodes.tsv\"  # Replace with your input file name\n",
    "output_file = \"10x_output_hESC/label.txt\"  # Output file name\n",
    "label = \"test cell type\"  # Desired label for all barcodes\n",
    "\n",
    "create_label_file(input_file, output_file, label)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2791ea1e-e976-4b53-a224-af6ebb395a2a",
   "metadata": {},
   "source": [
    "At this step we have converted `ExpressionData.csv` to pseudo 10x output with no accesibility data in it)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d6ff058-7ad6-4a9b-8a29-85972e03cbb8",
   "metadata": {},
   "source": [
    "We also need (or do we? - to be determined) to provide chromoseome locations for the genes in expression matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2df1dae6-ae45-4080-8977-dbc68496600c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unmatched genes saved to: unmatched_genes.tsv\n",
      "Updated features file created: 10x_output_hESC/updated_features_encode.tsv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "\n",
    "def extract_gene_locations_from_encode_gtf(gtf_file):\n",
    "    \"\"\"\n",
    "    Extract gene locations (chromosome, start, end) from an ENCODE GTF file.\n",
    "    \n",
    "    Args:\n",
    "        gtf_file (str): Path to the ENCODE GTF file.\n",
    "    \n",
    "    Returns:\n",
    "        dict: A dictionary mapping gene names to (chromosome, start, end).\n",
    "    \"\"\"\n",
    "    gene_locations = {}\n",
    "    with gzip.open(gtf_file, 'rt') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue  # Skip header lines\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            if fields[2] == \"gene\":  # Only process gene entries\n",
    "                info = {kv.split(\" \")[0]: kv.split(\" \")[1].strip('\";') for kv in fields[8].split(\"; \") if kv}\n",
    "                gene_name = info.get(\"gene_name\", None)\n",
    "                if gene_name:\n",
    "                    # Extract chromosome, start, and end\n",
    "                    chromosome = fields[0]\n",
    "                    start = int(fields[3])\n",
    "                    end = int(fields[4])\n",
    "                    gene_locations[gene_name] = (chromosome, start, end)\n",
    "    return gene_locations\n",
    "\n",
    "\n",
    "def update_features_with_locations(features_file, gtf_file, output_file, unmatched_file=None):\n",
    "    \"\"\"\n",
    "    Update features.tsv with chromosome, start, and end locations using an ENCODE GTF file.\n",
    "    \n",
    "    Args:\n",
    "        features_file (str): Path to the input features.tsv file.\n",
    "        gtf_file (str): Path to the ENCODE GTF file.\n",
    "        output_file (str): Path to save the updated features.tsv file.\n",
    "        unmatched_file (str, optional): Path to save unmatched genes for review.\n",
    "    \"\"\"\n",
    "    # Load the existing features file\n",
    "    features = pd.read_csv(features_file, sep=\"\\t\", header=None, names=[\"gene_id\", \"gene_name\", \"feature_type\"])\n",
    "\n",
    "    # Extract gene locations from the ENCODE GTF file\n",
    "    gtf_gene_locations = extract_gene_locations_from_encode_gtf(gtf_file)\n",
    "\n",
    "    # Merge the features data with gene locations from the GTF file\n",
    "    updated_features = features.copy()\n",
    "    updated_features[\"chromosome\"] = updated_features[\"gene_name\"].map(lambda gene_name: gtf_gene_locations.get(gene_name, (None, None, None))[0])\n",
    "    updated_features[\"start\"] = updated_features[\"gene_name\"].map(lambda gene_name: gtf_gene_locations.get(gene_name, (None, None, None))[1])\n",
    "    updated_features[\"end\"] = updated_features[\"gene_name\"].map(lambda gene_name: gtf_gene_locations.get(gene_name, (None, None, None))[2])\n",
    "\n",
    "    # Separate unmatched genes for logging if needed\n",
    "    unmatched_genes = updated_features[updated_features[\"chromosome\"].isna()]\n",
    "    if unmatched_file and not unmatched_genes.empty:\n",
    "        unmatched_genes[[\"gene_id\", \"gene_name\"]].to_csv(unmatched_file, sep=\"\\t\", index=False, header=False)\n",
    "        print(f\"Unmatched genes saved to: {unmatched_file}\")\n",
    "\n",
    "    # Save the updated features file (including unmatched rows)\n",
    "    updated_features.to_csv(output_file, sep=\"\\t\", index=False, header=False)\n",
    "    print(f\"Updated features file created: {output_file}\")\n",
    "    return gtf_gene_locations\n",
    "\n",
    "# Example usage\n",
    "unmatched_file = \"unmatched_genes.tsv\"  # Optional file to save unmatched genes\n",
    "features_file = \"10x_output_hESC/features.tsv\"  # Input features.tsv file\n",
    "gtf_file = \"gencode.v7.annotation.gtf.gz\"  # ENCODE GTF file\n",
    "output_file = \"10x_output_hESC/updated_features_encode.tsv\"  # Output file\n",
    "\n",
    "gene_locations = update_features_with_locations(features_file, gtf_file, output_file, unmatched_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "dfb081bd-2c09-4bf3-afca-f5442192cfe0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DDX11L1': ('chr1', 11869, 14409),\n",
       " 'AL627309.2': ('chr1', 11872, 14412),\n",
       " 'DDX11L11': ('chr9', 11987, 14522),\n",
       " 'WASH7P': ('chr1', 14363, 29806),\n",
       " 'MIR1302-10': ('chr19', 71973, 72110),\n",
       " 'FAM138A': ('chr1', 34554, 36081),\n",
       " 'OR4G11P': ('chr1', 62948, 63887),\n",
       " 'AL627309.1': ('chr1', 65882, 70008),\n",
       " 'RP11-34P13.7': ('chr1', 89295, 133566),\n",
       " 'RP11-34P13.8': ('chr1', 89551, 91105)}"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import islice\n",
    "subset = dict(islice(gene_locations.items(), 10))\n",
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "8f176ceb-bfb8-4b0e-a375-2f20bc3a2932",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_genes_in_annot = list(gene_locations.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "63fa6847-092d-411c-a919-688087b65e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_genes_in_annot.sort()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5b7aded3-1ae2-4eca-ae47-566a9f1d203a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "481062"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "Path(\"sorted_genes.txt\").write_text('\\n'.join(list_of_genes_in_annot))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "5c97fc7f-fba1-4bfd-8f41-40201a5c6bba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAED1', 'ABRACL', 'ACTL10', 'ADGB', 'ADTRP', 'AJUBA', 'AKIP1', 'ALG9',\n",
       "       'ALYREF', 'ANKHD1-EIF4EBP3',\n",
       "       ...\n",
       "       'ZNF322', 'ZNF559-ZNF177', 'ZNF587B', 'ZNF664-FAM101A', 'ZNF721',\n",
       "       'ZNF735', 'ZNF816-ZNF321P', 'ZNF850', 'ZNF853', 'ZNF865'],\n",
       "      dtype='object', length=737)"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_cell_type = pd.read_csv('GSE75748_sc_cell_type_ec.csv', index_col=0)\n",
    "original_cell_type.index.difference(list_of_genes_in_annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "257efa1a-83d5-4d55-bf1e-57535e353e43",
   "metadata": {},
   "source": [
    "## Step 1 - preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7a908d5-8256-492f-b9dd-b3fc2ad1def6",
   "metadata": {},
   "source": [
    "### Extraction of matrices for RNASeq and ATACSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "3a898b4f-9f03-4a8b-b3a1-4c2a3d19843c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/preprocess.py:278: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_RNA.obs['label']=label.loc[adata_RNA.obs['barcode']]['label'].values\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/preprocess.py:281: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_ATAC.obs['label']=label.loc[adata_ATAC.obs['barcode']]['label'].values\n"
     ]
    }
   ],
   "source": [
    "from LingerGRN.preprocess import *\n",
    "adata_RNA, adata_ATAC = get_adata(matrix, features, barcodes, label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "71d3d106-ae43-4d53-93b9-1b09e2c740e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 757 × 17735\n",
       "    obs: 'barcode', 'sample', 'label', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt'\n",
       "    var: 'gene_ids', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80fb676a-10f3-4d9a-a2f9-a82a00fbc65d",
   "metadata": {},
   "source": [
    "### Removing low count cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "cd9b1eab-c66e-45d1-90cf-ff6495df20e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata_RNA, min_genes=200)\n",
    "sc.pp.filter_genes(adata_RNA, min_cells=3)\n",
    "#sc.pp.filter_cells(adata_ATAC, min_genes=200)\n",
    "#sc.pp.filter_genes(adata_ATAC, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "78382d67-31f1-40d4-a027-80a28c6fca39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find common barcodes between RNA and ATAC datasets\n",
    "selected_barcode = list(set(adata_RNA.obs['barcode']) & set(adata_ATAC.obs['barcode']))\n",
    "\n",
    "# Filter both datasets to keep only the shared barcodes\n",
    "adata_RNA = adata_RNA[adata_RNA.obs['barcode'].isin(selected_barcode)]\n",
    "adata_ATAC = adata_ATAC[adata_ATAC.obs['barcode'].isin(selected_barcode)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d462dad-3136-4392-8f36-fa005027afcd",
   "metadata": {},
   "source": [
    "### Generating metacells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a894b0-2dcb-42ab-8ff8-98b747a5b81c",
   "metadata": {},
   "source": [
    "Metacells are generated to create more stable, representative profiles of groups of cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "22899e88-32f6-44b1-ba0b-d000b8f2d05b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LingerGRN.pseudo_bulk import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "a093f756-bea8-4b84-ab67-6a306d1bcffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplelist=list(set(adata_ATAC.obs['sample'].values)) \n",
    "tempsample=samplelist[0]\n",
    "TG_pseudobulk=pd.DataFrame([])\n",
    "RE_pseudobulk=pd.DataFrame([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "2538e4df-a7ee-487a-b1df-c6df6a86383f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_normalization.py:169: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:842: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:372: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Cannot cut empty array",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[106], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m adata_RNAtemp\u001b[38;5;241m=\u001b[39madata_RNA[adata_RNA\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msample\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39mtempsample]\n\u001b[1;32m      4\u001b[0m adata_ATACtemp\u001b[38;5;241m=\u001b[39madata_ATAC[adata_ATAC\u001b[38;5;241m.\u001b[39mobs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msample\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39mtempsample]\n\u001b[0;32m----> 5\u001b[0m TG_pseudobulk_temp,RE_pseudobulk_temp\u001b[38;5;241m=\u001b[39m\u001b[43mpseudo_bulk\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata_RNAtemp\u001b[49m\u001b[43m,\u001b[49m\u001b[43madata_ATACtemp\u001b[49m\u001b[43m,\u001b[49m\u001b[43msinglepseudobulk\u001b[49m\u001b[43m)\u001b[49m                \n\u001b[1;32m      6\u001b[0m TG_pseudobulk\u001b[38;5;241m=\u001b[39mpd\u001b[38;5;241m.\u001b[39mconcat([TG_pseudobulk, TG_pseudobulk_temp], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      7\u001b[0m RE_pseudobulk\u001b[38;5;241m=\u001b[39mpd\u001b[38;5;241m.\u001b[39mconcat([RE_pseudobulk, RE_pseudobulk_temp], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/pseudo_bulk.py:54\u001b[0m, in \u001b[0;36mpseudo_bulk\u001b[0;34m(adata_RNA, adata_ATAC, singlepseudobulk)\u001b[0m\n\u001b[1;32m     52\u001b[0m sc\u001b[38;5;241m.\u001b[39mpp\u001b[38;5;241m.\u001b[39mlog1p(adata_ATAC)\n\u001b[1;32m     53\u001b[0m sc\u001b[38;5;241m.\u001b[39mpp\u001b[38;5;241m.\u001b[39mfilter_genes(adata_ATAC, min_cells\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m) \n\u001b[0;32m---> 54\u001b[0m \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhighly_variable_genes\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata_ATAC\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmin_mean\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.0125\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_mean\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmin_disp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.5\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     55\u001b[0m adata_ATAC\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m=\u001b[39madata_ATAC\n\u001b[1;32m     56\u001b[0m adata_ATAC \u001b[38;5;241m=\u001b[39m adata_ATAC[:, adata_ATAC\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mhighly_variable]\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:439\u001b[0m, in \u001b[0;36mhighly_variable_genes\u001b[0;34m(adata, layer, n_top_genes, min_disp, max_disp, min_mean, max_mean, span, n_bins, flavor, subset, inplace, batch_key, check_values)\u001b[0m\n\u001b[1;32m    427\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _highly_variable_genes_seurat_v3(\n\u001b[1;32m    428\u001b[0m         adata,\n\u001b[1;32m    429\u001b[0m         layer\u001b[38;5;241m=\u001b[39mlayer,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    435\u001b[0m         inplace\u001b[38;5;241m=\u001b[39minplace,\n\u001b[1;32m    436\u001b[0m     )\n\u001b[1;32m    438\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m batch_key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 439\u001b[0m     df \u001b[38;5;241m=\u001b[39m \u001b[43m_highly_variable_genes_single_batch\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    440\u001b[0m \u001b[43m        \u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlayer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlayer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmin_disp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmin_disp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmax_disp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmax_disp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmin_mean\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmin_mean\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmax_mean\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmax_mean\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m        \u001b[49m\u001b[43mn_top_genes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_top_genes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m        \u001b[49m\u001b[43mn_bins\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_bins\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m        \u001b[49m\u001b[43mflavor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mflavor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    451\u001b[0m     sanitize_anndata(adata)\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:214\u001b[0m, in \u001b[0;36m_highly_variable_genes_single_batch\u001b[0;34m(adata, layer, min_disp, max_disp, min_mean, max_mean, n_top_genes, n_bins, flavor)\u001b[0m\n\u001b[1;32m    212\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdispersions\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m dispersion\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m flavor \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mseurat\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 214\u001b[0m     df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean_bin\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmeans\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbins\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_bins\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    215\u001b[0m     disp_grouped \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mgroupby(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean_bin\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdispersions\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    216\u001b[0m     disp_mean_bin \u001b[38;5;241m=\u001b[39m disp_grouped\u001b[38;5;241m.\u001b[39mmean()\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/pandas/core/reshape/tile.py:256\u001b[0m, in \u001b[0;36mcut\u001b[0;34m(x, bins, right, labels, retbins, precision, include_lowest, duplicates, ordered)\u001b[0m\n\u001b[1;32m    253\u001b[0m     sz \u001b[38;5;241m=\u001b[39m x\u001b[38;5;241m.\u001b[39msize\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m sz \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot cut empty array\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    258\u001b[0m rng \u001b[38;5;241m=\u001b[39m (nanops\u001b[38;5;241m.\u001b[39mnanmin(x), nanops\u001b[38;5;241m.\u001b[39mnanmax(x))\n\u001b[1;32m    259\u001b[0m mn, mx \u001b[38;5;241m=\u001b[39m (mi \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m0.0\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m mi \u001b[38;5;129;01min\u001b[39;00m rng)\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot cut empty array"
     ]
    }
   ],
   "source": [
    "singlepseudobulk = (adata_RNA.obs['sample'].unique().shape[0]*adata_RNA.obs['sample'].unique().shape[0]>100)\n",
    "for tempsample in samplelist:\n",
    "    adata_RNAtemp=adata_RNA[adata_RNA.obs['sample']==tempsample]\n",
    "    adata_ATACtemp=adata_ATAC[adata_ATAC.obs['sample']==tempsample]\n",
    "    TG_pseudobulk_temp,RE_pseudobulk_temp=pseudo_bulk(adata_RNAtemp,adata_ATACtemp,singlepseudobulk)                \n",
    "    TG_pseudobulk=pd.concat([TG_pseudobulk, TG_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk=pd.concat([RE_pseudobulk, RE_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk[RE_pseudobulk > 100] = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a362b22b-3f9a-45de-b42f-5a2bef81309b",
   "metadata": {},
   "source": [
    "Write preprocessed data to a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7370a90d-6e4c-45fa-856a-1f140c37cd46",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "linger_storage = 'results/linger_storage'\n",
    "os.makedirs(linger_storage, exist_ok = True)\n",
    "\n",
    "adata_ATAC.write(os.path.join(linger_storage,'adata_ATAC.h5ad'))\n",
    "adata_RNA.write(os.path.join(linger_storage,'adata_RNA.h5ad'))\n",
    "TG_pseudobulk=TG_pseudobulk.fillna(0)\n",
    "RE_pseudobulk=RE_pseudobulk.fillna(0)\n",
    "pd.DataFrame(adata_ATAC.var['gene_ids']).to_csv('data/Peaks.txt',header=None,index=None) #hard coded path to Peaks in preprocess\n",
    "TG_pseudobulk.to_csv(os.path.join(linger_storage,'TG_pseudobulk.tsv'))\n",
    "RE_pseudobulk.to_csv(os.path.join(linger_storage,'RE_pseudobulk.tsv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287022e0-b7a1-404a-9cab-4bd6a4d91c48",
   "metadata": {},
   "source": [
    "## Step 2 - training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7991602d-ae5e-4a04-a256-9efc18e5cbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = 'LINGER'\n",
    "Datadir = '/home/kl467102/proj_v0/downloads/' # Directory for the downloaded general gene regulatory network for LINGER\n",
    "GRNdir = Datadir+'data_bulk/'\n",
    "genome = 'hg38'\n",
    "outdir = os.path.join(os.getcwd(),'results/LINGER_output/')\n",
    "os.makedirs(outdir, exist_ok = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591a8b26-303b-4b68-8408-16a94d4e7e46",
   "metadata": {},
   "source": [
    "NOTE TO SELF: Because of `pybedtools` dependency `outdir` needs to be specified as an absolute path, otherwise resulting in an error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e359f0ef-1952-48ad-aa80-4c1530d7c889",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LingerGRN.preprocess import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6c512acc-5f74-4da2-8193-8f151504528b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping gene expression...\n",
      "Generate TF expression...\n",
      "Generate RE chromatin accessibility...\n",
      "Generate TF binding...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 23/23 [1:12:17<00:00, 188.60s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generate Index...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████| 14907/14907 [05:31<00:00, 44.97it/s]\n"
     ]
    }
   ],
   "source": [
    "preprocess(TG_pseudobulk,RE_pseudobulk,GRNdir,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e404e9c3-ad51-41e8-ab6e-21ad003f004b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)\n"
     ]
    }
   ],
   "source": [
    "import LingerGRN.LINGER_tr as LINGER_tr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4d36b18-e610-4815-893b-607a012649f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▌                                                                             | 12/1520 [09:21<19:35:47, 46.78s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m activef\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReLU\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mLINGER_tr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtraining\u001b[49m\u001b[43m(\u001b[49m\u001b[43mGRNdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43moutdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mactivef\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mHuman\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/LINGER_tr.py:414\u001b[0m, in \u001b[0;36mtraining\u001b[0;34m(GRNdir, method, outdir, activef, species)\u001b[0m\n\u001b[1;32m    412\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ii \u001b[38;5;129;01min\u001b[39;00m tqdm(\u001b[38;5;28mrange\u001b[39m(N)):\n\u001b[1;32m    413\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mfilterwarnings(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 414\u001b[0m     res\u001b[38;5;241m=\u001b[39m\u001b[43msc_nn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mii\u001b[49m\u001b[43m,\u001b[49m\u001b[43mgene_chr\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTFindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTFindex_bulk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex_bulk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex_bulk_match\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTarget\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnetall\u001b[49m\u001b[43m,\u001b[49m\u001b[43madj_matrix_all\u001b[49m\u001b[43m,\u001b[49m\u001b[43mExp\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTF_match\u001b[49m\u001b[43m,\u001b[49m\u001b[43minput_size_all\u001b[49m\u001b[43m,\u001b[49m\u001b[43mfisherall\u001b[49m\u001b[43m,\u001b[49m\u001b[43mOpn\u001b[49m\u001b[43m,\u001b[49m\u001b[43ml1_lambda\u001b[49m\u001b[43m,\u001b[49m\u001b[43mfisher_w\u001b[49m\u001b[43m,\u001b[49m\u001b[43mactivef\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    415\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mresetwarnings()\n\u001b[1;32m    416\u001b[0m     index_all\u001b[38;5;241m=\u001b[39mgene_chr\u001b[38;5;241m.\u001b[39mindex[ii]\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/LINGER_tr.py:177\u001b[0m, in \u001b[0;36msc_nn\u001b[0;34m(ii, gene_chr, TFindex, TFindex_bulk, REindex, REindex_bulk, REindex_bulk_match, Target, netall, adj_matrix_all, Exp, TF_match, input_size_all, fisherall, Opn, l1_lambda, fisher_w, activef)\u001b[0m\n\u001b[1;32m    175\u001b[0m     \u001b[38;5;66;03m# Perform backpropagation\u001b[39;00m\n\u001b[1;32m    176\u001b[0m     optimizer\u001b[38;5;241m.\u001b[39mzero_grad()\n\u001b[0;32m--> 177\u001b[0m     \u001b[43mloss\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m     optimizer\u001b[38;5;241m.\u001b[39mstep()\n\u001b[1;32m    179\u001b[0m np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mseed(\u001b[38;5;241m42\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/torch/_tensor.py:492\u001b[0m, in \u001b[0;36mTensor.backward\u001b[0;34m(self, gradient, retain_graph, create_graph, inputs)\u001b[0m\n\u001b[1;32m    482\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m has_torch_function_unary(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    483\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m handle_torch_function(\n\u001b[1;32m    484\u001b[0m         Tensor\u001b[38;5;241m.\u001b[39mbackward,\n\u001b[1;32m    485\u001b[0m         (\u001b[38;5;28mself\u001b[39m,),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    490\u001b[0m         inputs\u001b[38;5;241m=\u001b[39minputs,\n\u001b[1;32m    491\u001b[0m     )\n\u001b[0;32m--> 492\u001b[0m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mautograd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/torch/autograd/__init__.py:251\u001b[0m, in \u001b[0;36mbackward\u001b[0;34m(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)\u001b[0m\n\u001b[1;32m    246\u001b[0m     retain_graph \u001b[38;5;241m=\u001b[39m create_graph\n\u001b[1;32m    248\u001b[0m \u001b[38;5;66;03m# The reason we repeat the same comment below is that\u001b[39;00m\n\u001b[1;32m    249\u001b[0m \u001b[38;5;66;03m# some Python versions print out the first line of a multi-line function\u001b[39;00m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;66;03m# calls in the traceback and some print out the last line\u001b[39;00m\n\u001b[0;32m--> 251\u001b[0m \u001b[43mVariable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execution_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_backward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Calls into the C++ engine to run the backward pass\u001b[39;49;00m\n\u001b[1;32m    252\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtensors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgrad_tensors_\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    256\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mallow_unreachable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    258\u001b[0m \u001b[43m    \u001b[49m\u001b[43maccumulate_grad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    259\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "activef='ReLU'\n",
    "LINGER_tr.training(GRNdir,method,outdir,activef,'Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ef831a-5bd4-49bc-9aeb-ee9417f5be18",
   "metadata": {},
   "source": [
    "## Step 3 - population GNR inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc9d6ced-fc6f-4636-aa45-4e633e7bd232",
   "metadata": {},
   "outputs": [],
   "source": [
    "import LingerGRN.LL_net as LL_net\n",
    "LL_net.TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7250d63b-a66c-489c-969b-707a38067f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb42b81-a92c-46c6-8c1e-0635c160f948",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.trans_reg(GRNdir,method,outdir,genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581d75e4-62ba-4779-8b46-4c8cf9c86783",
   "metadata": {},
   "source": [
    "## Step 4 - cell-type specific GNR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0af7a02-1457-4a64-9039-3a606ec96a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype='all'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b416341e-ea6d-419c-bb90-9582ecadc739",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir,method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b98fa7cf-0c72-459b-ba2c-5dcad3bdedca",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b340b58-2e6e-4009-852d-4cc86658c92e",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_trans_reg(GRNdir,adata_RNA,celltype,outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6b44a6d-a5df-494f-a3d1-c9d897921c87",
   "metadata": {},
   "source": [
    "# Session information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e707da9b-eb4b-44d2-9d18-c2d7b2b6780f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "LingerGRN           NA\n",
       "anndata             0.9.2\n",
       "numpy               1.24.3\n",
       "pandas              2.0.3\n",
       "scanpy              1.9.5\n",
       "scipy               1.11.3\n",
       "session_info        1.0.0\n",
       "tqdm                4.67.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                         11.0.0\n",
       "anyio                       NA\n",
       "arrow                       1.3.0\n",
       "asttokens                   NA\n",
       "attr                        24.2.0\n",
       "attrs                       24.2.0\n",
       "babel                       2.16.0\n",
       "brotli                      1.1.0\n",
       "certifi                     2024.08.30\n",
       "cffi                        1.17.1\n",
       "charset_normalizer          3.4.0\n",
       "cloudpickle                 3.1.0\n",
       "colorama                    0.4.6\n",
       "comm                        0.2.2\n",
       "cycler                      0.12.1\n",
       "cython_runtime              NA\n",
       "dateutil                    2.9.0\n",
       "debugpy                     1.8.8\n",
       "decorator                   5.1.1\n",
       "defusedxml                  0.7.1\n",
       "exceptiongroup              1.2.2\n",
       "executing                   2.1.0\n",
       "fastjsonschema              NA\n",
       "fqdn                        NA\n",
       "h5py                        3.12.1\n",
       "idna                        3.10\n",
       "ipykernel                   6.29.5\n",
       "ipywidgets                  8.1.5\n",
       "isoduration                 NA\n",
       "jedi                        0.19.2\n",
       "jinja2                      3.1.4\n",
       "joblib                      1.3.2\n",
       "json5                       0.9.28\n",
       "jsonpointer                 3.0.0\n",
       "jsonschema                  4.23.0\n",
       "jsonschema_specifications   NA\n",
       "jupyter_events              0.10.0\n",
       "jupyter_server              2.14.2\n",
       "jupyterlab_server           2.27.3\n",
       "kiwisolver                  1.4.7\n",
       "llvmlite                    0.43.0\n",
       "markupsafe                  3.0.2\n",
       "matplotlib                  3.8.0\n",
       "matplotlib_inline           0.1.7\n",
       "mpl_toolkits                NA\n",
       "mpmath                      1.3.0\n",
       "natsort                     8.4.0\n",
       "nbformat                    5.10.4\n",
       "numba                       0.60.0\n",
       "overrides                   NA\n",
       "packaging                   24.2\n",
       "parso                       0.8.4\n",
       "pickleshare                 0.7.5\n",
       "platformdirs                4.3.6\n",
       "prometheus_client           NA\n",
       "prompt_toolkit              3.0.48\n",
       "psutil                      6.1.0\n",
       "pure_eval                   0.2.3\n",
       "pybedtools                  0.10.0\n",
       "pycparser                   2.22\n",
       "pydev_ipython               NA\n",
       "pydevconsole                NA\n",
       "pydevd                      3.2.2\n",
       "pydevd_file_utils           NA\n",
       "pydevd_plugins              NA\n",
       "pydevd_tracing              NA\n",
       "pygments                    2.18.0\n",
       "pynndescent                 0.5.13\n",
       "pyparsing                   3.2.0\n",
       "pysam                       0.22.1\n",
       "pythonjsonlogger            NA\n",
       "pytz                        2024.2\n",
       "referencing                 NA\n",
       "requests                    2.32.3\n",
       "rfc3339_validator           0.1.4\n",
       "rfc3986_validator           0.1.1\n",
       "rpds                        NA\n",
       "send2trash                  NA\n",
       "shap                        0.42.0\n",
       "six                         1.16.0\n",
       "sklearn                     1.3.0\n",
       "slicer                      NA\n",
       "sniffio                     1.3.1\n",
       "socks                       1.7.1\n",
       "stack_data                  0.6.2\n",
       "sympy                       1.13.3\n",
       "threadpoolctl               3.5.0\n",
       "torch                       2.1.0+cu121\n",
       "torchgen                    NA\n",
       "tornado                     6.4.1\n",
       "traitlets                   5.14.3\n",
       "typing_extensions           NA\n",
       "umap                        0.5.7\n",
       "uri_template                NA\n",
       "urllib3                     2.2.3\n",
       "wcwidth                     0.2.13\n",
       "webcolors                   24.8.0\n",
       "websocket                   1.8.0\n",
       "yaml                        6.0.2\n",
       "zmq                         26.2.0\n",
       "zoneinfo                    NA\n",
       "zstandard                   0.23.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.29.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.7.2\n",
       "jupyterlab          4.2.5\n",
       "notebook            7.2.2\n",
       "-----\n",
       "Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]\n",
       "Linux-6.8.0-x86_64-AMD_Opteron-tm-_Processor_6380-with-glibc2.39\n",
       "-----\n",
       "Session information updated at 2024-11-16 13:31\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04c92ec5-5614-4376-bafb-14b6e8e206aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "linger_enviroment",
   "language": "python",
   "name": "linger_enviroment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
