{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6bea81bb-5517-48a7-98f4-2c55eb7e0c1b",
   "metadata": {},
   "source": [
    "# LINGER ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00b25efc-3bd5-4e13-9714-6a68c7037295",
   "metadata": {},
   "source": [
    "In this notebook, we present a GNR inference, largely based on the works of:\n",
    "> [Yuan, Qiuyue, and Zhana Duren. \"Inferring gene regulatory networks from single-cell multiome data using atlas-scale external data.\" Nature Biotechnology (2024): 1-11.](https://doi.org/10.1038/s41587-024-02182-7)\n",
    "\n",
    "and the tutorial available at [LINGER's Github](https://github.com/Durenlab/LINGER/blob/main/docs/PBMC.md)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c058d3af-934c-457e-a38f-cb48918b2412",
   "metadata": {},
   "source": [
    "The data used are output files from ***PBMC from a Healthy Donor - Granulocytes Removed Through Cell Sorting (10k)*** *Single Cell Multiome ATAC + Gene Expression experiment* on *10X Chromium* platform analysed with Cell Ranger ARC, which are available on [10XGenomics website](https://www.10xgenomics.com/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-2-0-0). Cell type annotations were downloaded from [PBMC tutorial](https://github.com/Durenlab/LINGER/blob/main/docs/PBMC.md)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cc667cf-0870-4cd8-b15c-dd6ba8062ebb",
   "metadata": {},
   "source": [
    "## Preparation steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af4dd3e8-950e-4f2e-aaa9-eb499bf95c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "os.chdir('/home/kl467102/thesis/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e9e06c3-69b2-4de5-939f-af3488915cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scipy\n",
    "import pandas as pd\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e1d6f02-ed13-4dec-80b1-8963360c3852",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = 'data/filtered_feature_bc_matrix' # path to 10x output files\n",
    "label_file = 'data/PBMC_label.txt' # path to cell-type annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af75bf70-b734-4a9f-83a4-5b4db898f8a0",
   "metadata": {},
   "source": [
    "The input consists of `matrix.mtx`, `features.tsv` and `barcodes.tsv` and an annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d0da79b9-8c59-4199-b201-c9d3d4647198",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix=scipy.io.mmread(os.path.join(input_dir, 'matrix.mtx'))\n",
    "features=pd.read_csv(os.path.join(input_dir, 'features.tsv'), sep='\\t', header=None)\n",
    "barcodes=pd.read_csv(os.path.join(input_dir, 'barcodes.tsv'), sep='\\t', header=None)\n",
    "label=pd.read_csv(label_file, sep='\\t', header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52acc68-f1e0-4837-b271-9b06f6162871",
   "metadata": {},
   "source": [
    "We have loaded the data, now we can inspect them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8dd76a2d-876f-461b-bc60-d6fd5c995848",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(180488, 11898)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b471ef9-31e0-4850-bc6a-6a922b8fe916",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553</td>\n",
       "      <td>30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180486</th>\n",
       "      <td>KI270713.1:31340-32243</td>\n",
       "      <td>KI270713.1:31340-32243</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>31340</td>\n",
       "      <td>32243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180487</th>\n",
       "      <td>KI270713.1:36927-37836</td>\n",
       "      <td>KI270713.1:36927-37836</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>36927</td>\n",
       "      <td>37836</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0                       1                2  \\\n",
       "0              ENSG00000243485             MIR1302-2HG  Gene Expression   \n",
       "1              ENSG00000237613                 FAM138A  Gene Expression   \n",
       "180486  KI270713.1:31340-32243  KI270713.1:31340-32243            Peaks   \n",
       "180487  KI270713.1:36927-37836  KI270713.1:36927-37836            Peaks   \n",
       "\n",
       "                 3      4      5  \n",
       "0             chr1  29553  30267  \n",
       "1             chr1  36080  36081  \n",
       "180486  KI270713.1  31340  32243  \n",
       "180487  KI270713.1  36927  37836  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([features.head(2), features.tail(2)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "39dc8531-8d7f-4db9-b428-a75ad0c6ccea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAGCCAAGGAATC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACAGCCAATCCCTT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCCAATGCGCT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0\n",
       "0  AAACAGCCAAGGAATC-1\n",
       "1  AAACAGCCAATCCCTT-1\n",
       "2  AAACAGCCAATGCGCT-1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "49a27e9c-94e3-4e9d-86e0-0286d5946698",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode_use</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAGCCAAGGAATC-1</td>\n",
       "      <td>naive CD4 T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACAGCCAATCCCTT-1</td>\n",
       "      <td>memory CD4 T cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCCAATGCGCT-1</td>\n",
       "      <td>naive CD4 T cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          barcode_use               label\n",
       "0  AAACAGCCAAGGAATC-1   naive CD4 T cells\n",
       "1  AAACAGCCAATCCCTT-1  memory CD4 T cells\n",
       "2  AAACAGCCAATGCGCT-1   naive CD4 T cells"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "257efa1a-83d5-4d55-bf1e-57535e353e43",
   "metadata": {},
   "source": [
    "## Step 1 - preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7a908d5-8256-492f-b9dd-b3fc2ad1def6",
   "metadata": {},
   "source": [
    "### Extraction of matrices for RNASeq and ATACSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3a898b4f-9f03-4a8b-b3a1-4c2a3d19843c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/preprocess.py:278: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_RNA.obs['label']=label.loc[adata_RNA.obs['barcode']]['label'].values\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/preprocess.py:281: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_ATAC.obs['label']=label.loc[adata_ATAC.obs['barcode']]['label'].values\n"
     ]
    }
   ],
   "source": [
    "from LingerGRN.preprocess import *\n",
    "adata_RNA, adata_ATAC = get_adata(matrix, features, barcodes, label)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80fb676a-10f3-4d9a-a2f9-a82a00fbc65d",
   "metadata": {},
   "source": [
    "### Removing low count cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cd9b1eab-c66e-45d1-90cf-ff6495df20e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:139: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['n_genes'] = number\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata_RNA, min_genes=200)\n",
    "sc.pp.filter_genes(adata_RNA, min_cells=3)\n",
    "sc.pp.filter_cells(adata_ATAC, min_genes=200)\n",
    "sc.pp.filter_genes(adata_ATAC, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "78382d67-31f1-40d4-a027-80a28c6fca39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find common barcodes between RNA and ATAC datasets\n",
    "selected_barcode = list(set(adata_RNA.obs['barcode']) & set(adata_ATAC.obs['barcode']))\n",
    "\n",
    "# Filter both datasets to keep only the shared barcodes\n",
    "adata_RNA = adata_RNA[adata_RNA.obs['barcode'].isin(selected_barcode)]\n",
    "adata_ATAC = adata_ATAC[adata_ATAC.obs['barcode'].isin(selected_barcode)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d462dad-3136-4392-8f36-fa005027afcd",
   "metadata": {},
   "source": [
    "### Generating metacells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a894b0-2dcb-42ab-8ff8-98b747a5b81c",
   "metadata": {},
   "source": [
    "Metacells are generated to create more stable, representative profiles of groups of cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22899e88-32f6-44b1-ba0b-d000b8f2d05b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LingerGRN.pseudo_bulk import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a093f756-bea8-4b84-ab67-6a306d1bcffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplelist=list(set(adata_ATAC.obs['sample'].values)) \n",
    "tempsample=samplelist[0]\n",
    "TG_pseudobulk=pd.DataFrame([])\n",
    "RE_pseudobulk=pd.DataFrame([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2538e4df-a7ee-487a-b1df-c6df6a86383f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_normalization.py:169: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:842: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:372: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:842: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    }
   ],
   "source": [
    "singlepseudobulk = (adata_RNA.obs['sample'].unique().shape[0]*adata_RNA.obs['sample'].unique().shape[0]>100)\n",
    "for tempsample in samplelist:\n",
    "    adata_RNAtemp=adata_RNA[adata_RNA.obs['sample']==tempsample]\n",
    "    adata_ATACtemp=adata_ATAC[adata_ATAC.obs['sample']==tempsample]\n",
    "    TG_pseudobulk_temp,RE_pseudobulk_temp=pseudo_bulk(adata_RNAtemp,adata_ATACtemp,singlepseudobulk)                \n",
    "    TG_pseudobulk=pd.concat([TG_pseudobulk, TG_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk=pd.concat([RE_pseudobulk, RE_pseudobulk_temp], axis=1)\n",
    "    RE_pseudobulk[RE_pseudobulk > 100] = 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a362b22b-3f9a-45de-b42f-5a2bef81309b",
   "metadata": {},
   "source": [
    "Write preprocessed data to a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7370a90d-6e4c-45fa-856a-1f140c37cd46",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/home/kl467102/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/anndata/_core/anndata.py:1230: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "linger_storage = 'results/linger_storage'\n",
    "os.makedirs(linger_storage, exist_ok = True)\n",
    "\n",
    "adata_ATAC.write(os.path.join(linger_storage,'adata_ATAC.h5ad'))\n",
    "adata_RNA.write(os.path.join(linger_storage,'adata_RNA.h5ad'))\n",
    "TG_pseudobulk=TG_pseudobulk.fillna(0)\n",
    "RE_pseudobulk=RE_pseudobulk.fillna(0)\n",
    "pd.DataFrame(adata_ATAC.var['gene_ids']).to_csv('data/Peaks.txt',header=None,index=None) #hard coded path to Peaks in preprocess\n",
    "TG_pseudobulk.to_csv(os.path.join(linger_storage,'TG_pseudobulk.tsv'))\n",
    "RE_pseudobulk.to_csv(os.path.join(linger_storage,'RE_pseudobulk.tsv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287022e0-b7a1-404a-9cab-4bd6a4d91c48",
   "metadata": {},
   "source": [
    "## Step 2 - training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7991602d-ae5e-4a04-a256-9efc18e5cbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = 'LINGER'\n",
    "Datadir = '/home/kl467102/proj_v0/downloads/' # Directory for the downloaded general gene regulatory network for LINGER\n",
    "GRNdir = Datadir+'data_bulk/'\n",
    "genome = 'hg38'\n",
    "outdir = os.path.join(os.getcwd(),'results/LINGER_output/')\n",
    "os.makedirs(outdir, exist_ok = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591a8b26-303b-4b68-8408-16a94d4e7e46",
   "metadata": {},
   "source": [
    "NOTE TO SELF: Because of `pybedtools` dependency `outdir` needs to be specified as an absolute path, otherwise resulting in an error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e359f0ef-1952-48ad-aa80-4c1530d7c889",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LingerGRN.preprocess import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6c512acc-5f74-4da2-8193-8f151504528b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping gene expression...\n",
      "Generate TF expression...\n",
      "Generate RE chromatin accessibility...\n",
      "Generate TF binding...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 23/23 [1:12:17<00:00, 188.60s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generate Index...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████| 14907/14907 [05:31<00:00, 44.97it/s]\n"
     ]
    }
   ],
   "source": [
    "preprocess(TG_pseudobulk,RE_pseudobulk,GRNdir,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e404e9c3-ad51-41e8-ab6e-21ad003f004b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)\n"
     ]
    }
   ],
   "source": [
    "import LingerGRN.LINGER_tr as LINGER_tr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4d36b18-e610-4815-893b-607a012649f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▌                                                                             | 12/1520 [09:21<19:35:47, 46.78s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m activef\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReLU\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mLINGER_tr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtraining\u001b[49m\u001b[43m(\u001b[49m\u001b[43mGRNdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43moutdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mactivef\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mHuman\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/LINGER_tr.py:414\u001b[0m, in \u001b[0;36mtraining\u001b[0;34m(GRNdir, method, outdir, activef, species)\u001b[0m\n\u001b[1;32m    412\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ii \u001b[38;5;129;01min\u001b[39;00m tqdm(\u001b[38;5;28mrange\u001b[39m(N)):\n\u001b[1;32m    413\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mfilterwarnings(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 414\u001b[0m     res\u001b[38;5;241m=\u001b[39m\u001b[43msc_nn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mii\u001b[49m\u001b[43m,\u001b[49m\u001b[43mgene_chr\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTFindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTFindex_bulk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex_bulk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mREindex_bulk_match\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTarget\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnetall\u001b[49m\u001b[43m,\u001b[49m\u001b[43madj_matrix_all\u001b[49m\u001b[43m,\u001b[49m\u001b[43mExp\u001b[49m\u001b[43m,\u001b[49m\u001b[43mTF_match\u001b[49m\u001b[43m,\u001b[49m\u001b[43minput_size_all\u001b[49m\u001b[43m,\u001b[49m\u001b[43mfisherall\u001b[49m\u001b[43m,\u001b[49m\u001b[43mOpn\u001b[49m\u001b[43m,\u001b[49m\u001b[43ml1_lambda\u001b[49m\u001b[43m,\u001b[49m\u001b[43mfisher_w\u001b[49m\u001b[43m,\u001b[49m\u001b[43mactivef\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    415\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mresetwarnings()\n\u001b[1;32m    416\u001b[0m     index_all\u001b[38;5;241m=\u001b[39mgene_chr\u001b[38;5;241m.\u001b[39mindex[ii]\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/LingerGRN/LINGER_tr.py:177\u001b[0m, in \u001b[0;36msc_nn\u001b[0;34m(ii, gene_chr, TFindex, TFindex_bulk, REindex, REindex_bulk, REindex_bulk_match, Target, netall, adj_matrix_all, Exp, TF_match, input_size_all, fisherall, Opn, l1_lambda, fisher_w, activef)\u001b[0m\n\u001b[1;32m    175\u001b[0m     \u001b[38;5;66;03m# Perform backpropagation\u001b[39;00m\n\u001b[1;32m    176\u001b[0m     optimizer\u001b[38;5;241m.\u001b[39mzero_grad()\n\u001b[0;32m--> 177\u001b[0m     \u001b[43mloss\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m     optimizer\u001b[38;5;241m.\u001b[39mstep()\n\u001b[1;32m    179\u001b[0m np\u001b[38;5;241m.\u001b[39mrandom\u001b[38;5;241m.\u001b[39mseed(\u001b[38;5;241m42\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/torch/_tensor.py:492\u001b[0m, in \u001b[0;36mTensor.backward\u001b[0;34m(self, gradient, retain_graph, create_graph, inputs)\u001b[0m\n\u001b[1;32m    482\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m has_torch_function_unary(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    483\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m handle_torch_function(\n\u001b[1;32m    484\u001b[0m         Tensor\u001b[38;5;241m.\u001b[39mbackward,\n\u001b[1;32m    485\u001b[0m         (\u001b[38;5;28mself\u001b[39m,),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    490\u001b[0m         inputs\u001b[38;5;241m=\u001b[39minputs,\n\u001b[1;32m    491\u001b[0m     )\n\u001b[0;32m--> 492\u001b[0m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mautograd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbackward\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/linger_enviroment/lib/python3.10/site-packages/torch/autograd/__init__.py:251\u001b[0m, in \u001b[0;36mbackward\u001b[0;34m(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)\u001b[0m\n\u001b[1;32m    246\u001b[0m     retain_graph \u001b[38;5;241m=\u001b[39m create_graph\n\u001b[1;32m    248\u001b[0m \u001b[38;5;66;03m# The reason we repeat the same comment below is that\u001b[39;00m\n\u001b[1;32m    249\u001b[0m \u001b[38;5;66;03m# some Python versions print out the first line of a multi-line function\u001b[39;00m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;66;03m# calls in the traceback and some print out the last line\u001b[39;00m\n\u001b[0;32m--> 251\u001b[0m \u001b[43mVariable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execution_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_backward\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Calls into the C++ engine to run the backward pass\u001b[39;49;00m\n\u001b[1;32m    252\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtensors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgrad_tensors_\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretain_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcreate_graph\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    256\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mallow_unreachable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    258\u001b[0m \u001b[43m    \u001b[49m\u001b[43maccumulate_grad\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    259\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "activef='ReLU'\n",
    "LINGER_tr.training(GRNdir,method,outdir,activef,'Human')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ef831a-5bd4-49bc-9aeb-ee9417f5be18",
   "metadata": {},
   "source": [
    "## Step 3 - population GNR inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc9d6ced-fc6f-4636-aa45-4e633e7bd232",
   "metadata": {},
   "outputs": [],
   "source": [
    "import LingerGRN.LL_net as LL_net\n",
    "LL_net.TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7250d63b-a66c-489c-969b-707a38067f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb42b81-a92c-46c6-8c1e-0635c160f948",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.trans_reg(GRNdir,method,outdir,genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581d75e4-62ba-4779-8b46-4c8cf9c86783",
   "metadata": {},
   "source": [
    "## Step 4 - cell-type specific GNR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0af7a02-1457-4a64-9039-3a606ec96a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype='all'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b416341e-ea6d-419c-bb90-9582ecadc739",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir,method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b98fa7cf-0c72-459b-ba2c-5dcad3bdedca",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b340b58-2e6e-4009-852d-4cc86658c92e",
   "metadata": {},
   "outputs": [],
   "source": [
    "LL_net.cell_type_specific_trans_reg(GRNdir,adata_RNA,celltype,outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6b44a6d-a5df-494f-a3d1-c9d897921c87",
   "metadata": {},
   "source": [
    "# Session information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e707da9b-eb4b-44d2-9d18-c2d7b2b6780f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "LingerGRN           NA\n",
       "anndata             0.9.2\n",
       "numpy               1.24.3\n",
       "pandas              2.0.3\n",
       "scanpy              1.9.5\n",
       "scipy               1.11.3\n",
       "session_info        1.0.0\n",
       "tqdm                4.67.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                         11.0.0\n",
       "anyio                       NA\n",
       "arrow                       1.3.0\n",
       "asttokens                   NA\n",
       "attr                        24.2.0\n",
       "attrs                       24.2.0\n",
       "babel                       2.16.0\n",
       "brotli                      1.1.0\n",
       "certifi                     2024.08.30\n",
       "cffi                        1.17.1\n",
       "charset_normalizer          3.4.0\n",
       "cloudpickle                 3.1.0\n",
       "colorama                    0.4.6\n",
       "comm                        0.2.2\n",
       "cycler                      0.12.1\n",
       "cython_runtime              NA\n",
       "dateutil                    2.9.0\n",
       "debugpy                     1.8.8\n",
       "decorator                   5.1.1\n",
       "defusedxml                  0.7.1\n",
       "exceptiongroup              1.2.2\n",
       "executing                   2.1.0\n",
       "fastjsonschema              NA\n",
       "fqdn                        NA\n",
       "h5py                        3.12.1\n",
       "idna                        3.10\n",
       "ipykernel                   6.29.5\n",
       "ipywidgets                  8.1.5\n",
       "isoduration                 NA\n",
       "jedi                        0.19.2\n",
       "jinja2                      3.1.4\n",
       "joblib                      1.3.2\n",
       "json5                       0.9.28\n",
       "jsonpointer                 3.0.0\n",
       "jsonschema                  4.23.0\n",
       "jsonschema_specifications   NA\n",
       "jupyter_events              0.10.0\n",
       "jupyter_server              2.14.2\n",
       "jupyterlab_server           2.27.3\n",
       "kiwisolver                  1.4.7\n",
       "llvmlite                    0.43.0\n",
       "markupsafe                  3.0.2\n",
       "matplotlib                  3.8.0\n",
       "matplotlib_inline           0.1.7\n",
       "mpl_toolkits                NA\n",
       "mpmath                      1.3.0\n",
       "natsort                     8.4.0\n",
       "nbformat                    5.10.4\n",
       "numba                       0.60.0\n",
       "overrides                   NA\n",
       "packaging                   24.2\n",
       "parso                       0.8.4\n",
       "pickleshare                 0.7.5\n",
       "platformdirs                4.3.6\n",
       "prometheus_client           NA\n",
       "prompt_toolkit              3.0.48\n",
       "psutil                      6.1.0\n",
       "pure_eval                   0.2.3\n",
       "pybedtools                  0.10.0\n",
       "pycparser                   2.22\n",
       "pydev_ipython               NA\n",
       "pydevconsole                NA\n",
       "pydevd                      3.2.2\n",
       "pydevd_file_utils           NA\n",
       "pydevd_plugins              NA\n",
       "pydevd_tracing              NA\n",
       "pygments                    2.18.0\n",
       "pynndescent                 0.5.13\n",
       "pyparsing                   3.2.0\n",
       "pysam                       0.22.1\n",
       "pythonjsonlogger            NA\n",
       "pytz                        2024.2\n",
       "referencing                 NA\n",
       "requests                    2.32.3\n",
       "rfc3339_validator           0.1.4\n",
       "rfc3986_validator           0.1.1\n",
       "rpds                        NA\n",
       "send2trash                  NA\n",
       "shap                        0.42.0\n",
       "six                         1.16.0\n",
       "sklearn                     1.3.0\n",
       "slicer                      NA\n",
       "sniffio                     1.3.1\n",
       "socks                       1.7.1\n",
       "stack_data                  0.6.2\n",
       "sympy                       1.13.3\n",
       "threadpoolctl               3.5.0\n",
       "torch                       2.1.0+cu121\n",
       "torchgen                    NA\n",
       "tornado                     6.4.1\n",
       "traitlets                   5.14.3\n",
       "typing_extensions           NA\n",
       "umap                        0.5.7\n",
       "uri_template                NA\n",
       "urllib3                     2.2.3\n",
       "wcwidth                     0.2.13\n",
       "webcolors                   24.8.0\n",
       "websocket                   1.8.0\n",
       "yaml                        6.0.2\n",
       "zmq                         26.2.0\n",
       "zoneinfo                    NA\n",
       "zstandard                   0.23.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.29.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.7.2\n",
       "jupyterlab          4.2.5\n",
       "notebook            7.2.2\n",
       "-----\n",
       "Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]\n",
       "Linux-6.8.0-x86_64-AMD_Opteron-tm-_Processor_6380-with-glibc2.39\n",
       "-----\n",
       "Session information updated at 2024-11-16 13:31\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04c92ec5-5614-4376-bafb-14b6e8e206aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "linger_enviroment",
   "language": "python",
   "name": "linger_enviroment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
